solutions/usaco-899.mdx

```diff
@@ -0,0 +1,113 @@
+---
+id: usaco-899
+source: USACO Gold 2019 January
+title: Shortcut
+author: Ryan Chou
+---
+
+[Official Analysis (C++)](http://www.usaco.org/current/data/sol_shortcut_gold_jan19.html)
+
+## Explanation
+
+A simple brute-force approach would be to find the shortest path for all of the nodes, and then try to add a shortcut from $1$ to $i$ and find the shortest path again for all $i$.
+
+Using this approach, the answer is $\min(\sum\limits_{i=0}^n \texttt{dist}-\texttt{short}_i)$, where $\texttt{dist}$ is the travel time that all cows take without a shortcut and $\texttt{short}_i$ represents the travel time that all cows take when there is a shortcut from $1$ to $i$. Although this will certainly TLE, it hints at a faster way to compute the time saved. 
+
+Let's call $\texttt{occ}_i$ the number of cows that pass through field $i$. Then the answer would be $\max(\sum\limits_{i=0}^n \texttt{occ}_i \cdot (\texttt{trav}_i - T))$. $\texttt{trav}_i$ being the cost for one cow to get from field $i$ to $1$.
+
+We can find the values for $\texttt{trav}_i$ quickly with [Djikstra's Algorithm](/gold/shortest-paths). To get $\texttt{occ}_i$, we'll store the parents of nodes during our Djikstra's Algorithm, so we can backtrack to get the counts of cows that pass through. This can be found in $\mathcal{O}(N^2)$ because $N \leq 10^4$.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(M\log N + N^2)$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+int main() {
+	freopen("shortcut.in", "r", stdin);
+	freopen("shortcut.out", "w", stdout);
+
+	int n, m, t;
+	cin >> n >> m >> t;
+
+	vector<int> fields(n);
+	for (int i = 0; i < n; i++) {
+		cin >> fields[i];
+	}
+
+	// adj[i] = {travel time, adjacent node}
+	vector<vector<pair<int, int>>> adj(n);
+	for (int i = 0; i < m; i++) {
+		int u, v, c;
+		cin >> u >> v >> c;
+
+		adj[--u].push_back({c, --v});
+		adj[v].push_back({c, u});
+	}
+
+	vector<int> cost(n, INT32_MAX);
+
+	// prev stores parents for backtracking
+	vector<int> prev(n, INT32_MAX);
+	priority_queue<
+		pair<int, int>, 
+		vector<pair<int, int>>, 
+		greater<pair<int, int>>
+	> pq;
+	
+	cost[0] = 0;
+	
+	// travel time, next node
```

Comment:
shouldn't next node be current onde instead?

Reply:
err i think it depends on the way you look at it, but ill change it to current node